prime-router/src/main/kotlin/reportfilehistory/ReportFileHistory.kt

```diff
@@ -0,0 +1,282 @@
+        warnings.addAll(consolidateLogs(ActionLogLevel.warning))
+    }
+
+    /**
+     * Consolidate the [logs] filtered by an optional [filterBy] action level, so to list similar messages once
+     * with a list of items they relate to.
+     * @return the consolidated list of logs
+     */
+    internal fun consolidateLogs(filterBy: ActionLogLevel? = null):
+        List<ConsolidatedActionLog> {
+        val consolidatedList = mutableListOf<ConsolidatedActionLog>()
+
+        // First filter the logs and sort by the message.  This first sorting can take care of sorting old messages
+        // that contain index numbers like "Report 3: xxxx"
+        val filteredList = when (filterBy) {
+            null -> logs
+            else -> logs.filter { it.type == filterBy }
+        }.sortedBy { it.detail.message }
+        // Now order the list so that logs contain first non-item messages, then item messages, and item messages
+        // are sorted by index.
+        val orderedList = (
+            filteredList.filter { it.scope != ActionLogScope.item }.sortedBy { it.scope } +
+                filteredList.filter { it.scope == ActionLogScope.item }.sortedBy { it.index }
+            ).toMutableList()
+        // Now consolidate the list
+        while (orderedList.isNotEmpty()) {
+            // Grab the first log.
+            val consolidatedLog = ConsolidatedActionLog(orderedList.first())
+            orderedList.removeFirst()
+            // Now find similar logs and consolidate it.  Note by using the iterator we can delete on the fly.
+            with(orderedList.iterator()) {
+                forEach { log ->
+                    if (consolidatedLog.canBeConsolidatedWith(log)) {
+                        consolidatedLog.add(log)
+                        remove()
+                    }
+                }
+            }
+            consolidatedList.add(consolidatedLog)
+        }
+
+        return consolidatedList
+    }
+}
+
+/**
+ * This is a container for various bits of report data used by DetailedReportFileHistory
+ *
+ * @property reportId unique identifier for this specific report
+ * @property receivingOrg where is this report going?
+ * @property receivingOrgSvc what service is receiving this report?
+ * @property sendingOrg who sent this report?
+ * @property sendingOrgClient what service did the sender use to send this report?
+ * @property schemaTopic the kind of data contained in the report (e.g. "covid-19")
+ * @property externalName actual filename of the report's file
+ * @property createdAt when the report was created
+ * @property nextActionAt when the report is next expected to send or process
+ * @property itemCount number of tests (data rows) contained in the report
+ * @property itemCountBeforeQualFilter number of tests that were submitted by the sender
+ */
+@JsonInclude(Include.NON_NULL)
+@JsonIgnoreProperties(ignoreUnknown = true)
+data class DetailReport(
```

Comment:
Please ensure consistency of 'Detail' or 'Detailed' in all classes. Doesn't matter which one, but please be consistent.